ligo/src/token/fa2.mligo

```diff
@@ -26,27 +22,46 @@ let check_sender (from_ , store : address * storage): address =
         ("FA2_NOT_OPERATOR", ()) : address)
 
 [@inline]
-let debit_from (amt, from_, token_id, ledger: nat * address * token_id * ledger): ledger =
+let debit_from (amt, from_, token_id, ledger, total_supply: nat * address * token_id * ledger * total_supply): (ledger * total_supply) =
   match Big_map.find_opt (from_, token_id) ledger with
     Some bal ->
-      begin
-        match Michelson.is_nat (bal - amt) with
-          Some new_bal -> Big_map.update (from_, token_id) (Some new_bal) ledger
+      let store =
+        match Map.find_opt token_id total_supply with
+          Some current_total_supply ->
+            (match Michelson.is_nat (current_total_supply - amt) with
+              Some new_total_supply ->
+                (match Michelson.is_nat (bal - amt) with
+                  Some new_bal ->
+                    let ledger = Big_map.update (from_, token_id) (Some new_bal) ledger in
+                    let total_supply = Map.update token_id (Some new_total_supply) total_supply in
+                    (ledger, total_supply)
+                | None ->
+                    ([%Michelson ({| { FAILWITH } |} : string * (nat * nat) -> (ledger * total_supply))] ("FA2_INSUFFICIENT_BALANCE", (amt, bal)) : (ledger * total_supply))
+                )
+            | None ->
+                (failwith("NEGATIVE_TOTAL_SUPPLY") : (ledger * total_supply))
```

Comment:
I think this should be splitted in steps, like:
- first `current_total_supply = ...` (put 0 in case of `None`)
- then `new_total_supply = ...` (fail on `None`)
- etc.

instead of indenting it so much.